client.go

```diff
@@ -135,45 +135,70 @@ type RequestParams struct {
 
 // Generic helper to execute HTTP requests using the net/http package
 //
-// nolint:unused
+// nolint:unused, gocognit
 func (c *httpClient) doRequest(ctx context.Context, method, url string, params RequestParams, mutators ...func(req *http.Request) error) error {
-	req, bodyBuffer, err := c.createRequest(ctx, method, url, params)
-	if err != nil {
-		return err
-	}
+	var (
+		req        *http.Request
+		bodyBuffer *bytes.Buffer
+		resp       *http.Response
+		err        error
+	)
 
-	if err := c.applyMutators(req, mutators); err != nil {
-		return err
-	}
+	for attempt := 0; attempt < httpDefaultRetryCount; attempt++ {
+		req, bodyBuffer, err = c.createRequest(ctx, method, url, params)
+		if err != nil {
+			return err
+		}
 
-	if c.debug && c.logger != nil {
-		c.logRequest(req, method, url, bodyBuffer)
-	}
+		if err = c.applyMutators(req, mutators); err != nil {
+			return err
+		}
 
-	resp, err := c.sendRequest(req)
-	if err != nil {
-		return err
-	}
-	defer resp.Body.Close()
+		if c.debug && c.logger != nil {
+			c.logRequest(req, method, url, bodyBuffer)
+		}
 
-	if err := c.checkHTTPError(resp); err != nil {
-		return err
-	}
+		resp, err = c.sendRequest(req)
+		if err == nil { //nolint:nestif
+			defer resp.Body.Close()
+			if err = c.checkHTTPError(resp); err == nil {
+				if c.debug && c.logger != nil {
+					resp, err = c.logResponse(resp)
+					if err != nil {
+						return err
+					}
+				}
+				if params.Response != nil {
+					if err = c.decodeResponseBody(resp, params.Response); err != nil {
+						return err
+					}
+				}
+				return nil
+			}
+		}
 
-	if c.debug && c.logger != nil {
-		resp, err = c.logResponse(resp)
-		if err != nil {
-			return err
+		if !c.shouldRetry(resp, err) {
+			break
 		}
-	}
 
-	if params.Response != nil {
-		if err := c.decodeResponseBody(resp, params.Response); err != nil {
-			return err
+		retryAfter, retryErr := c.retryAfter(resp)
+		if retryErr != nil {
+			return retryErr
 		}
+		time.Sleep(retryAfter)
```

Comment:
I was a bit confused on this part until I found that 0 is returned by `c.retryAfter(...)` if the `Retry-After` header isn't found. I think that's a good design decision but it might be handy to have a comment clarifying that

Reply:
Done!